lib/manual_relocator.rb

```diff
@@ -0,0 +1,352 @@
+  end
+
+private
+
+  def fetch_manual(slug)
+    manuals = ManualRecord.where(slug: slug)
+    raise "No manual found for slug '#{slug}'" if manuals.count == 0
+    raise "More than one manual found for slug '#{slug}'" if manuals.count > 1
+    manuals.first
+  end
+
+  def old_manual_document_ids
+    @old_manual_document_ids ||= old_manual.editions.flat_map(&:document_ids).uniq
+  end
+
+  def new_manual_document_ids
+    @new_manual_document_ids ||= new_manual.editions.flat_map(&:document_ids).uniq
+  end
+
+  def validate_manuals
+    raise "Manual to remove (#{to_slug}) should be published" unless manual_is_currently_published?(old_manual)
+    raise "Manual to reslug (#{from_slug}) should be published" unless manual_is_currently_published?(new_manual)
+  end
+
+  def manual_is_currently_published?(manual)
+    # to be currently published either...
+    # 1. it's got one edition that is published
+    (manual.editions.count == 1 && manual.latest_edition.state == "published") ||
+    # or
+    # 2. the last two editions are published and draft
+    (manual.editions.order_by([:version_number, :desc]).limit(2).map(&:state) == %w(draft published))
+  end
+
+  def redirect_and_remove
+    if old_manual.editions.any?
+      # Redirect all sections of the manual we're going to remove
+      # to prevent dead bookmarked URLs.
+      old_manual_document_ids.each do |document_id|
+        editions = all_editions_of_section(document_id)
+        section_slug = editions.first.slug
+
+        begin
+          if old_sections_reused_in_new_manual.include? document_id
+            puts "Issuing gone for content item '/#{section_slug}' as it will be reused by a section in '#{new_manual.slug}'"
+            send_gone(document_id, section_slug)
+          else
+            puts "Redirecting content item '/#{section_slug}' to '/#{old_manual.slug}'"
+            publishing_api.unpublish(document_id,
+                                     type: "redirect",
+                                     alternative_path: "/#{old_manual.slug}",
+                                     discard_drafts: true)
+          end
+        rescue GdsApi::HTTPNotFound
+          puts "Content item with content_id #{document_id} not present in the publishing API"
+        end
+
+        # Destroy all the editons of this manual as it's going away
+        editions.map(&:destroy)
+      end
+    end
+
+    puts "Destroying old PublicationLogs for #{old_manual.slug}"
+    PublicationLog.change_notes_for(old_manual.slug).each { |log| log.destroy }
+
+    # Destroy the manual record
+    puts "Destroying manual #{old_manual.manual_id}"
+    old_manual.destroy
+
+    puts "Issuing gone for #{old_manual.manual_id}"
+    send_gone(old_manual.manual_id, old_manual.slug)
+  end
+
+  def old_sections_reused_in_new_manual
+    @old_sections_reused_in_new_manual ||= _calculate_old_sections_reused_in_new_manual
+  end
+
+  def _calculate_old_sections_reused_in_new_manual
```

Comment:
What's the intent (presumably a convention) of method names beginning with an underscore? I'm asking out of plain old ignorance.